- !pip install -U sentence-transformers

In [1]:
%%time
from wh_qa.wh_qa import *

Wall time: 59.8 s


In [2]:
import dill as pickle
with open("test_questions/a1_questions.dill", "rb") as dill_file:
    data = pickle.load(dill_file)
print(len(data))

111


In [8]:
set_data = set()
for element in data:
    set_data.add(element['question'])
set_data

{'During what years did this prevent the normal flooding of the Nile?',
 'How long would Egyptian art last?',
 'How many pyramids did Pharaoh build?',
 'How many pyramids did Snefru build?',
 'How many years of famine and strife did the collapse of the Old Kingdom last?',
 'Ineb-Hedg was the Ancient Egyptian name for what?',
 'Local dynasties were largely independent from what authority?',
 'The Old Kingdom is most commonly regarded as the period from the Third Dynasty through to what?',
 'What Dynasties are in Egypt?',
 'What are the first lifesize statues created by sculptors?',
 'What art was perfected by the art of carving?',
 'What could have caused his death?',
 'What country traded cedar with?',
 'What did Egyptians focus on?',
 'What did Lebanon trade with Lebanon for?',
 'What did Pharaoh believe was necessary for crops?',
 'What did ship builders use to keep their ships assembled?',
 'What did the 4th dynasty focus on?',
 'What did the dynasties become?',
 'What did they perc

In [10]:
textfile = open("questions.txt", "w")
for element in set_data:
    textfile.write(element+ "\n")
textfile.close()

In [2]:
%%time
import dill as pickle
with open("test_questions/a1_questions.dill", "rb") as dill_file:
    data = pickle.load(dill_file)
print(len(data))

passage_embedding, passages, sentences = get_embs("test_questions/a1.txt")

accuracy = 0
count = 0
for i in range(len(data)):
    try:
        question = data[i]['question']
        ans = get_wh_answer(sentences, question, passage_embedding, passages, 1)
        print("Question ", count + 1,": Pred Ans - ", ans, " Correct Ans - ", data[i]['answer'])
        p = get_position(ans, data[i]['answer'])
        if p != 0:
            accuracy += 1/p
        count += 1
    except:
        pass
    
print(accuracy, " of ", count,": " ,accuracy/count)

111
Question  1 : Pred Ans -  ["the old kingdom is the period in the third millennium ( c. 2686 - 2181 bc ) also known as the'age of the pyramids"]  Correct Ans -  the great 4th Dynasty
Question  2 : Pred Ans -  ["the old kingdom is the period in the third millennium ( c. 2686 - 2181 bc ) also known as the'age of the pyramids"]  Correct Ans -  the great 4th Dynasty
Question  3 : Pred Ans -  ["the old kingdom is the period in the third millennium ( c. 2686 - 2181 bc ) also known as the'age of the pyramids"]  Correct Ans -  the great 4th Dynasty
Question  4 : Pred Ans -  ['middle kingdom and new kingdom']  Correct Ans -  the Middle Kingdom and New Kingdom
Question  5 : Pred Ans -  ['lower nile valley']  Correct Ans -  the lower Nile Valley
Question  6 : Pred Ans -  ['middle kingdom and new kingdom']  Correct Ans -  the Middle Kingdom and New Kingdom
Question  7 : Pred Ans -  ['the old kingdom and the early dynastic period']  Correct Ans -  the Old Kingdom and the Early Dynastic Period
Qu

Question  105 : Pred Ans -  ['relief decoration']  Correct Ans -  intricate relief decoration
Question  106 : Pred Ans -  ['to ensure an ordered existence and to defeat death']  Correct Ans -  to ensure an ordered existence and to defeat death
Question  107 : Pred Ans -  ['to ensure an ordered existence and to defeat death']  Correct Ans -  to ensure an ordered existence
Question  108 : Pred Ans -  ['more than 3, 000 years']  Correct Ans -  3,000 years
Question  109 : Pred Ans -  ['flexible']  Correct Ans -  flexible
Question  110 : Pred Ans -  ['preserving life after death']  Correct Ans -  preserving life after death
Question  111 : Pred Ans -  ['elegant']  Correct Ans -  jewelry
81.0  of  111 :  0.7297297297297297
Wall time: 21.9 s


In [3]:
with open("test_questions/a1_self_gen_qsts.dill", "rb") as dill_file:
    data = pickle.load(dill_file)
print(len(data))

224


In [4]:
for q in data:
    print(q)
    print(get_wh_answer(sentences, q, passage_embedding, passages, 1))

What is  the period in the third millennium (c. 2686-2181 BC) also known as the 'Age of the Pyramids' or 'Age of the Pyramid Builders' as it includes the great 4th Dynasty when King Sneferu perfected the art of pyramid building and the pyramids of Giza ?
['4th']
During what era did the king of Egypt become a living god?
['the old kingdom']
During what dynasty were pyramid complexes fewer devoted to construction than during the 4th?
['4th']
For what reason did artists create images and forms that endured generations?
["to express their culture's worldview"]
What is the period in the third millennium also known as the 'Age of the Pyramids'?
['Old Kingdom']
During what dynasty did the power of pharaoh gradually wane?
['Sixth Dynasty']
What been proposed  to be the work of Khafra and Khufu himself ?
['the sphinx']
What is  the revolutionary change in architecture accompanied by the effects on Egyptian society and economy of large-scale building projects ?
['effects on egyptian society']
Wh

['giza']
What believing  that he ensured the annual flooding of the Nile that was necessary for their crops ?
['pharaoh']
What was reached  However at Saqqara  ?
["the building of'the great pyramids"]
What was initiated  at Saqqara under his reign  ?
['a new era of building']
What was followed  by decades of famine and strife  ?
['the collapse of the old kingdom']
What began  with Userkaf (2494–2487 BC)  ?
['the fifth dynasty']
What were constructed  under the kings Khufu, Khafre, and Menkaure  ?
['giza']
Who was succeeded  by his son Sahure (2487–2475 BC) who commanded an expedition to Punt  ?
['userkaf']
What was coined  by eighteenth-century historians and the distinction between the Old Kingdom and the Early Dynastic Period  ?
['the Old Kingdom']
What were  Menkauhor Kaiu (2421–2414 BC), Djedkare Isesi (2414–2375 BC) and Unas (2375–2345), the earliest ruler to have the pyramid texts inscribed in his pyramid ?
['djedkare isesi ( 2414 – 2375 bc ) and unas']
What remained  at Ineb-Hed

- a1.txt: 0.73 (top 1), 0.80 (top 10)
- a2.txt: 0.80 (top 1),  0.83 (top 10)
- a3.txt: 0.73 (top 1),  0.78 (top 10)
- a4.txt: 0.77 (top 1),  0.80 (top 10)
- a5.txt: 0.77 (top 1),  0.82 (top 10)

import dill as pickle
with open("questions2", "rb") as dill_file:
    questions = pickle.load(dill_file)
len(questions)

f = open("a2.txt", encoding = "utf-8", mode = "r")
txt = f.read()
txt = txt[txt.find("\n\n\n")+3:txt.find("References")]

sentences = enumerate(sent_tokenize(txt))
passages = [[str(index), sentence] for index, sentence in sentences]
passage_embedding = bi_encoder.encode(passages)

%%time
accuracy = 0
count = 0

for i in range(len(data)):
    try: 
        question = data[i][0]['question']
        correct_answer = data[i][0]['answer']
        pred_answers =  get_wh_answer2(question, passage_embedding, passages, 10)   
        pos = get_position(pred_answers, correct_answer)
        if pos != 0:
                accuracy += 1/pos
        #print(count, " position: ", pos)
        #print()
        count += 1
    except:
        pass

accuracy, accuracy/count, count